### ISSUE TITLE
Random Forest sampling ratio

### ISSUE THREAD

**Issue by CreateRandom** [haifengl/smile#161]
Hi there,

another question regarding the Random Forest: comments in the code seem to indicate that for values of `subsample < 1.0`, samples are drawn _without_ replacement. I am a little confused about the semantics of the parameter in question: From the literature (compare [[Martínez-Muñoz & Suárez 2010](https://repositorio.uam.es/bitstream/handle/10486/664127/out-of-bag_martinez-munoz_PR_2010_ps.pdf?sequence=1)]), it appears to me that the sampling ratio only describes the size of the bagged subsets as a fraction of the original set size. That would mean that e.g. setting `subsample = 0.2` would not in itself imply that sampling _without_ replacement would have to take place. Similar to the authors of the paper cited, I would like to test different sampling ratios (while always sampling _with_ replacement). Is this possible in SMILE? Thanks!

**Comment by haifengl** [comment]
In general, random forest samples data by bootstrap (i.e. sample the data WITH replacement n times). Here is an extension the allows sampling without replacement. But without replacement, we have to specify a lower sample size. Otherwise, we will always get the same original samples.

**Comment by CreateRandom** [comment]
So, which sampling method would be used if I were to specify `subsample = 0.8`?

**Comment by haifengl** [comment]
Sampling WITHOUT replacement. Only subsample =1.0 means sampling with replacement.

**Comment by CreateRandom** [comment]
Sorry to keep bothering you about this, I just don't see the logic to it yet: Suppose I had a training set with n = 100 examples and would now want each tree in the RF to be trained on bagged subsets containing 80 examples each (i.e. I would set `subsample` to 0.8). Now, wouldn't it be possible to draw 80 examples _with_ replacement? This is essentially what I'm trying to do and I would say what many users are trying to do when setting `subsample` to values smaller than 1.0. Especially since the authors of the paper mentioned above suggest that for sampling with replacement, `subsample = 0.8` is usually a good option. Thanks again in advance!

**Comment by haifengl** [comment]
It is just a convention. We keep the same practice for random forest and gradient boosting.

**Comment by CreateRandom** [comment]
Alright, I kept thinking I was missing something. Thanks for letting me know. Presumably, if I want sampling with replacement with smaller subset sizes, I will just make the necessary adjustments to the code on my own fork.

**Comment by haifengl** [comment]
In fact, our sampling without replacement is more advanced as we consider the unbalanced distribution among classes. I think it is better than sampling a smaller size with replacement .
